File: src/RouteManager.cpp
#include "amr_fleet_adapter/RouteManager.hpp"
#include <rmf_traffic/agv/Graph.hpp>
#include <fstream>
#include <yaml-cpp/yaml.h>
#include <opencv2/opencv.hpp>

namespace amr_fleet_adapter
{

RouteManager::RouteManager(rclcpp::Node::SharedPtr node)
: node_(node)
, map_loaded_(false)
{
  RCLCPP_INFO(node_->get_logger(), "Creating RouteManager");
}

RouteManager::~RouteManager() = default;

bool RouteManager::initialize()
{
  RCLCPP_INFO(node_->get_logger(), "Initializing RouteManager");
  
  // Create map publisher for visualization
  map_pub_ = node_->create_publisher<nav_msgs::msg::OccupancyGrid>("/map", 1);
  
  // Create timer to publish map periodically
  map_publish_timer_ = node_->create_wall_timer(
    std::chrono::seconds(5),
    [this]() { publish_map(); });
  
  return true;
}

bool RouteManager::load_routes_from_json(const std::string& json_file_path)
{
  RCLCPP_INFO(node_->get_logger(), "Loading routes from JSON file: %s", json_file_path.c_str());
  
  try
  {
    std::ifstream file(json_file_path);
    if (!file.is_open())
    {
      RCLCPP_ERROR(node_->get_logger(), "Cannot open JSON file: %s", json_file_path.c_str());
      return false;
    }
    
    nlohmann::json json_data;
    file >> json_data;
    file.close();
    
    return parse_json_routes(json_data);
  }
  catch (const std::exception& e)
  {
    RCLCPP_ERROR(node_->get_logger(), "Exception loading routes from JSON: %s", e.what());
    return false;
  }
}

bool RouteManager::load_routes_from_json_string(const std::string& json_string)
{
  RCLCPP_INFO(node_->get_logger(), "Loading routes from JSON string");
  
  try
  {
    nlohmann::json json_data = nlohmann::json::parse(json_string);
    return parse_json_routes(json_data);
  }
  catch (const std::exception& e)
  {
    RCLCPP_ERROR(node_->get_logger(), "Exception parsing JSON string: %s", e.what());
    return false;
  }
}

bool RouteManager::load_map_info(const std::string& map_yaml_file)
{
  RCLCPP_INFO(node_->get_logger(), "Loading map info from: %s", map_yaml_file.c_str());
  
  try
  {
    YAML::Node config = YAML::LoadFile(map_yaml_file);
    
    map_info_.yaml_file_path = map_yaml_file;
    map_info_.resolution = config["resolution"].as<double>();
    map_info_.map_name = config["image"].as<std::string>();
    
    // Handle relative path for PGM file
    std::string pgm_path = config["image"].as<std::string>();
    if (pgm_path[0] != '/')
    {
      // Relative path - combine with YAML file directory
      std::string yaml_dir = map_yaml_file.substr(0, map_yaml_file.find_last_of("/\\"));
      map_info_.pgm_file_path = yaml_dir + "/" + pgm_path;
    }
    else
    {
      map_info_.pgm_file_path = pgm_path;
    }
    
    // Get origin
    auto origin = config["origin"].as<std::vector<double>>();
    if (origin.size() >= 2)
    {
      map_info_.origin[0] = origin[0];
      map_info_.origin[1] = origin[1];
    }
    
    // Load the PGM map
    if (!load_pgm_map(map_info_.pgm_file_path))
    {
      RCLCPP_ERROR(node_->get_logger(), "Failed to load PGM map: %s", map_info_.pgm_file_path.c_str());
      return false;
    }
    
    map_loaded_ = true;
    RCLCPP_INFO(node_->get_logger(), "Map loaded successfully: %s", map_info_.map_name.c_str());
    return true;
  }
  catch (const std::exception& e)
  {
    RCLCPP_ERROR(node_->get_logger(), "Exception loading map info: %s", e.what());
    return false;
  }
}

rmf_traffic::agv::Graph RouteManager::create_rmf_graph() const
{
  RCLCPP_INFO(node_->get_logger(), "Creating RMF traffic graph from routes");
  
  rmf_traffic::agv::Graph graph;
  
  // Create waypoints first
  std::unordered_map<std::string, std::size_t> waypoint_indices;
  
  for (const auto& [waypoint_name, waypoint] : waypoints_)
  {
    std::size_t index = graph.add_waypoint(waypoint->map_name, {waypoint->x, waypoint->y});
    waypoint_indices[waypoint_name] = index;
    
    RCLCPP_DEBUG(node_->get_logger(), "Added waypoint '%s' at index %zu [%.2f, %.2f]",
                 waypoint_name.c_str(), index, waypoint->x, waypoint->y);
  }
  
  // Create lanes between connected waypoints
  for (const auto& [waypoint_name, waypoint] : waypoints_)
  {
    std::size_t from_index = waypoint_indices[waypoint_name];
    
    for (const auto& connection : waypoint->connections)
    {
      auto it = waypoint_indices.find(connection);
      if (it != waypoint_indices.end())
      {
        std::size_t to_index = it->second;
        
        // Add bidirectional lane
        graph.add_lane(from_index, to_index);
        
        RCLCPP_DEBUG(node_->get_logger(), "Added lane from '%s' (%zu) to '%s' (%zu)",
                     waypoint_name.c_str(), from_index, connection.c_str(), to_index);
      }
      else
      {
        RCLCPP_WARN(node_->get_logger(), "Connection '%s' not found for waypoint '%s'",
                    connection.c_str(), waypoint_name.c_str());
      }
    }
  }
  
  RCLCPP_INFO(node_->get_logger(), "Created RMF graph with %zu waypoints and %zu lanes",
              graph.num_waypoints(), graph.num_lanes());
  
  return graph;
}

std::shared_ptr<Route> RouteManager::get_route(const std::string& route_id) const
{
  auto it = routes_.find(route_id);
  return (it != routes_.end()) ? it->second : nullptr;
}

std::vector<std::shared_ptr<Route>> RouteManager::get_all_routes() const
{
  std::vector<std::shared_ptr<Route>> all_routes;
  all_routes.reserve(routes_.size());
  
  for (const auto& [id, route] : routes_)
  {
    all_routes.push_back(route);
  }
  
  return all_routes;
}

std::shared_ptr<Route> RouteManager::find_route(
  const std::string& start_waypoint,
  const std::string& end_waypoint) const
{
  for (const auto& [id, route] : routes_)
  {
    if (route->start_waypoint == start_waypoint && route->end_waypoint == end_waypoint)
    {
      return route;
    }
  }
  return nullptr;
}

std::shared_ptr<RouteWaypoint> RouteManager::get_waypoint(const std::string& waypoint_name) const
{
  auto it = waypoints_.find(waypoint_name);
  return (it != waypoints_.end()) ? it->second : nullptr;
}

geometry_msgs::msg::PoseStamped RouteManager::waypoint_to_pose(const RouteWaypoint& waypoint) const
{
  geometry_msgs::msg::PoseStamped pose;
  pose.header.frame_id = "map";
  pose.header.stamp = node_->get_clock()->now();
  
  pose.pose.position.x = waypoint.x;
  pose.pose.position.y = waypoint.y;
  pose.pose.position.z = 0.0;
  
  // Convert theta to quaternion
  tf2::Quaternion q;
  q.setRPY(0, 0, waypoint.theta);
  pose.pose.orientation = tf2::toMsg(q);
  
  return pose;
}

std::shared_ptr<RouteWaypoint> RouteManager::find_nearest_waypoint(
  const geometry_msgs::msg::PoseStamped& pose,
  double max_distance) const
{
  std::shared_ptr<RouteWaypoint> nearest_waypoint = nullptr;
  double min_distance = max_distance;
  
  for (const auto& [name, waypoint] : waypoints_)
  {
    double dx = pose.pose.position.x - waypoint->x;
    double dy = pose.pose.position.y - waypoint->y;
    double distance = std::sqrt(dx*dx + dy*dy);
    
    if (distance < min_distance)
    {
      min_distance = distance;
      nearest_waypoint = waypoint;
    }
  }
  
  return nearest_waypoint;
}

const MapInfo& RouteManager::get_map_info() const
{
  return map_info_;
}

geometry_msgs::msg::Point RouteManager::map_to_world(double map_x, double map_y) const
{
  geometry_msgs::msg::Point world_point;
  
  if (map_loaded_)
  {
    world_point.x = map_info_.origin[0] + map_x * map_info_.resolution;
    world_point.y = map_info_.origin[1] + map_y * map_info_.resolution;
    world_point.z = 0.0;
  }
  else
  {
    world_point.x = map_x;
    world_point.y = map_y;
    world_point.z = 0.0;
  }
  
  return world_point;
}

std::pair<double, double> RouteManager::world_to_map(const geometry_msgs::msg::Point& world_point) const
{
  if (map_loaded_)
  {
    double map_x = (world_point.x - map_info_.origin[0]) / map_info_.resolution;
    double map_y = (world_point.y - map_info_.origin[1]) / map_info_.resolution;
    return {map_x, map_y};
  }
  else
  {
    return {world_point.x, world_point.y};
  }
}

bool RouteManager::validate_routes() const
{
  RCLCPP_INFO(node_->get_logger(), "Validating routes...");
  
  bool valid = true;
  
  // Validate waypoint connections
  if (!validate_waypoint_connections())
  {
    valid = false;
  }
  
  // Validate individual routes
  for (const auto& [id, route] : routes_)
  {
    if (!validate_route_waypoints(*route))
    {
      valid = false;
    }
  }
  
  RCLCPP_INFO(node_->get_logger(), "Route validation %s", valid ? "passed" : "failed");
  return valid;
}

std::string RouteManager::export_routes_to_json() const
{
  nlohmann::json json_data;
  
  // Export routes
  json_data["routes"] = nlohmann::json::array();
  for (const auto& [id, route] : routes_)
  {
    json_data["routes"].push_back(*route);
  }
  
  // Export waypoints
  json_data["waypoints"] = nlohmann::json::array();
  for (const auto& [name, waypoint] : waypoints_)
  {
    json_data["waypoints"].push_back(*waypoint);
  }
  
  return json_data.dump(2);
}

bool RouteManager::parse_json_routes(const nlohmann::json& json_data)
{
  try
  {
    // Clear existing data
    routes_.clear();
    waypoints_.clear();
    
    // Parse waypoints first
    if (json_data.contains("waypoints"))
    {
      for (const auto& wp_json : json_data["waypoints"])
      {
        RouteWaypoint waypoint = parse_waypoint(wp_json);
        waypoints_[waypoint.name] = std::make_shared<RouteWaypoint>(waypoint);
        
        RCLCPP_DEBUG(node_->get_logger(), "Parsed waypoint: %s [%.2f, %.2f]",
                     waypoint.name.c_str(), waypoint.x, waypoint.y);
      }
    }
    
    // Parse routes
    if (json_data.contains("routes"))
    {
      for (const auto& route_json : json_data["routes"])
      {
        Route route = parse_route(route_json);
        routes_[route.route_id] = std::make_shared<Route>(route);
        
        RCLCPP_DEBUG(node_->get_logger(), "Parsed route: %s (%s -> %s)",
                     route.name.c_str(), route.start_waypoint.c_str(), route.end_waypoint.c_str());
      }
    }
    
    RCLCPP_INFO(node_->get_logger(), "Loaded %zu waypoints and %zu routes",
                waypoints_.size(), routes_.size());
    
    return validate_routes();
  }
  catch (const std::exception& e)
  {
    RCLCPP_ERROR(node_->get_logger(), "Exception parsing JSON routes: %s", e.what());
    return false;
  }
}

bool RouteManager::load_pgm_map(const std::string& pgm_file_path)
{
  try
  {
    // Load image using OpenCV
    cv::Mat map_image = cv::imread(pgm_file_path, cv::IMREAD_GRAYSCALE);
    
    if (map_image.empty())
    {
      RCLCPP_ERROR(node_->get_logger(), "Could not load PGM file: %s", pgm_file_path.c_str());
      return false;
    }
    
    // Convert to OccupancyGrid
    map_info_.occupancy_grid.header.frame_id = "map";
    map_info_.occupancy_grid.header.stamp = node_->get_clock()->now();
    
    map_info_.occupancy_grid.info.resolution = map_info_.resolution;
    map_info_.occupancy_grid.info.width = map_image.cols;
    map_info_.occupancy_grid.info.height = map_image.rows;
    
    map_info_.occupancy_grid.info.origin.position.x = map_info_.origin[0];
    map_info_.occupancy_grid.info.origin.position.y = map_info_.origin[1];
    map_info_.occupancy_grid.info.origin.position.z = 0.0;
    map_info_.occupancy_grid.info.origin.orientation.w = 1.0;
    
    // Convert image data to occupancy grid
    map_info_.occupancy_grid.data.resize(map_image.rows * map_image.cols);
    
    for (int y = 0; y < map_image.rows; ++y)
    {
      for (int x = 0; x < map_image.cols; ++x)
      {
        int index = (map_image.rows - 1 - y) * map_image.cols + x; // Flip Y axis
        uint8_t pixel = map_image.at<uint8_t>(y, x);
        
        // Convert pixel value to occupancy probability
        if (pixel > 250)
        {
          map_info_.occupancy_grid.data[index] = 0;    // Free space
        }
        else if (pixel < 5)
        {
          map_info_.occupancy_grid.data[index] = 100;  // Occupied
        }
        else
        {
          map_info_.occupancy_grid.data[index] = -1;   // Unknown
        }
      }
    }
    
    RCLCPP_INFO(node_->get_logger(), "Loaded PGM map: %dx%d, resolution: %.3f",
                map_image.cols, map_image.rows, map_info_.resolution);
    
    return true;
  }
  catch (const std::exception& e)
  {
    RCLCPP_ERROR(node_->get_logger(), "Exception loading PGM map: %s", e.what());
    return false;
  }
}

RouteWaypoint RouteManager::parse_waypoint(const nlohmann::json& waypoint_json)
{
  RouteWaypoint waypoint;
  
  waypoint.name = waypoint_json["name"].get<std::string>();
  waypoint.x = waypoint_json["x"].get<double>();
  waypoint.y = waypoint_json["y"].get<double>();
  waypoint.theta = waypoint_json.value("theta", 0.0);
  waypoint.map_name = waypoint_json.value("map_name", "map");
  
  if (waypoint_json.contains("connections"))
  {
    waypoint.connections = waypoint_json["connections"].get<std::vector<std::string>>();
  }
  
  return waypoint;
}

Route RouteManager::parse_route(const nlohmann::json& route_json)
{
  Route route;
  
  route.route_id = route_json["route_id"].get<std::string>();
  route.name = route_json["name"].get<std::string>();
  route.description = route_json.value("description", "");
  route.start_waypoint = route_json["start_waypoint"].get<std::string>();
  route.end_waypoint = route_json["end_waypoint"].get<std::string>();
  
  if (route_json.contains("waypoints"))
  {
    for (const auto& wp_json : route_json["waypoints"])
    {
      route.waypoints.push_back(parse_waypoint(wp_json));
    }
  }
  
  route.estimated_duration = calculate_route_duration(route);
  
  return route;
}

double RouteManager::calculate_route_duration(const Route& route) const
{
  // Simple estimation based on distance and average speed
  double total_distance = 0.0;
  const double average_speed = 0.5; // m/s
  
  if (route.waypoints.size() < 2)
  {
    return 30.0; // Default 30 seconds
  }
  
  for (size_t i = 0; i < route.waypoints.size() - 1; ++i)
  {
    const auto& wp1 = route.waypoints[i];
    const auto& wp2 = route.waypoints[i + 1];
    
    double dx = wp2.x - wp1.x;
    double dy = wp2.y - wp1.y;
    total_distance += std::sqrt(dx*dx + dy*dy);
  }
  
  return total_distance / average_speed;
}

void RouteManager::publish_map()
{
  if (map_loaded_ && map_pub_->get_subscription_count() > 0)
  {
    map_info_.occupancy_grid.header.stamp = node_->get_clock()->now();
    map_pub_->publish(map_info_.occupancy_grid);
  }
}

bool RouteManager::validate_waypoint_connections() const
{
  bool valid = true;
  
  for (const auto& [name, waypoint] : waypoints_)
  {
    for (const auto& connection : waypoint->connections)
    {
      if (waypoints_.find(connection) == waypoints_.end())
      {
        RCLCPP_ERROR(node_->get_logger(), 
                     "Waypoint '%s' has invalid connection to '%s'",
                     name.c_str(), connection.c_str());
        valid = false;
      }
    }
  }
  
  return valid;
}

bool RouteManager::validate_route_waypoints(const Route& route) const
{
  bool valid = true;
  
  // Check if start and end waypoints exist
  if (waypoints_.find(route.start_waypoint) == waypoints_.end())
  {
    RCLCPP_ERROR(node_->get_logger(), 
                 "Route '%s' has invalid start waypoint: '%s'",
                 route.name.c_str(), route.start_waypoint.c_str());
    valid = false;
  }
  
  if (waypoints_.find(route.end_waypoint) == waypoints_.end())
  {
    RCLCPP_ERROR(node_->get_logger(), 
                 "Route '%s' has invalid end waypoint: '%s'",
                 route.name.c_str(), route.end_waypoint.c_str());
    valid = false;
  }
  
  return valid;
}

// JSON serialization implementations
void to_json(nlohmann::json& j, const RouteWaypoint& waypoint)
{
  j = nlohmann::json{
    {"name", waypoint.name},
    {"x", waypoint.x},
    {"y", waypoint.y},
    {"theta", waypoint.theta},
    {"map_name", waypoint.map_name},
    {"connections", waypoint.connections}
  };
}

void from_json(const nlohmann::json& j, RouteWaypoint& waypoint)
{
  j.at("name").get_to(waypoint.name);
  j.at("x").get_to(waypoint.x);
  j.at("y").get_to(waypoint.y);
  waypoint.theta = j.value("theta", 0.0);
  waypoint.map_name = j.value("map_name", "map");
  
  if (j.contains("connections"))
  {
    j.at("connections").get_to(waypoint.connections);
  }
}

void to_json(nlohmann::json& j, const Route& route)
{
  j = nlohmann::json{
    {"route_id", route.route_id},
    {"name", route.name},
    {"description", route.description},
    {"waypoints", route.waypoints},
    {"start_waypoint", route.start_waypoint},
    {"end_waypoint", route.end_waypoint},
    {"estimated_duration", route.estimated_duration}
  };
}

void from_json(const nlohmann::json& j, Route& route)
{
  j.at("route_id").get_to(route.route_id);
  j.at("name").get_to(route.name);
  route.description = j.value("description", "");
  j.at("start_waypoint").get_to(route.start_waypoint);
  j.at("end_waypoint").get_to(route.end_waypoint);
  
  if (j.contains("waypoints"))
  {
    j.at("waypoints").get_to(route.waypoints);
  }
  
  route.estimated_duration = j.value("estimated_duration", 30.0);
}

} // namespace amr_fleet_adapter

File: amr_fleet_adapter/route_manager.py
"""
Route Manager Module

Handles route and map management for the AMR fleet adapter.
Loads JSON routes and PGM maps for navigation planning.
"""

import json
import yaml
import math
from typing import Dict, List, Optional, Any, Tuple
from dataclasses import dataclass
from datetime import datetime

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from geometry_msgs.msg import PoseStamped, Point, Quaternion
from nav_msgs.msg import OccupancyGrid
import cv2
import numpy as np


@dataclass
class Waypoint:
    """Waypoint definition."""
    name: str
    x: float
    y: float
    theta: float = 0.0
    map_name: str = "map"
    connections: List[str] = None
    
    def __post_init__(self):
        if self.connections is None:
            self.connections = []


@dataclass
class Route:
    """Route definition."""
    route_id: str
    name: str
    description: str = ""
    waypoints: List[Waypoint] = None
    start_waypoint: str = ""
    end_waypoint: str = ""
    estimated_duration: float = 0.0
    
    def __post_init__(self):
        if self.waypoints is None:
            self.waypoints = []


class RouteManager(Node):
    """
    Manages routes and maps for the AMR fleet.
    
    Loads waypoints and routes from JSON files,
    handles map loading from PGM/YAML files.
    """
    
    def __init__(self, parent_node: Node, node_name: str = "route_manager"):
        super().__init__(node_name)
        
        self.parent_node = parent_node
        self.get_logger().info("Initializing Route Manager...")
        
        # Storage
        self.waypoints: Dict[str, Waypoint] = {}
        self.routes: Dict[str, Route] = {}
        
        # Map information
        self.map_info = {
            'resolution': 0.05,
            'origin': [0.0, 0.0, 0.0],
            'width': 0,
            'height': 0,
            'map_name': '',
            'loaded': False
        }
        self.occupancy_grid: Optional[OccupancyGrid] = None
        
        # Publishers
        self.map_pub = self.create_publisher(OccupancyGrid, '/map', 1)
        
        # Timer for map publishing
        self.map_timer = self.create_timer(5.0, self._publish_map)
        
        self.get_logger().info("Route Manager initialized")
    
    def load_routes_from_json(self, json_file_path: str) -> bool:
        """Load routes from JSON file."""
        try:
            self.get_logger().info(f"Loading routes from: {json_file_path}")
            
            with open(json_file_path, 'r') as f:
                data = json.load(f)
            
            # Load waypoints
            if 'waypoints' in data:
                for wp_data in data['waypoints']:
                    waypoint = Waypoint(
                        name=wp_data['name'],
                        x=wp_data['x'],
                        y=wp_data['y'],
                        theta=wp_data.get('theta', 0.0),
                        map_name=wp_data.get('map_name', 'map'),
                        connections=wp_data.get('connections', [])
                    )
                    self.waypoints[waypoint.name] = waypoint
                    
                self.get_logger().info(f"Loaded {len(self.waypoints)} waypoints")
            
            # Load routes
            if 'routes' in data:
                for route_data in data['routes']:
                    route = Route(
                        route_id=route_data['route_id'],
                        name=route_data['name'],
                        description=route_data.get('description', ''),
                        start_waypoint=route_data.get('start_waypoint', ''),
                        end_waypoint=route_data.get('end_waypoint', ''),
                        estimated_duration=route_data.get('estimated_duration', 0.0)
                    )
                    
                    # Load route waypoints if specified
                    if 'waypoints' in route_data:
                        for wp_data in route_data['waypoints']:
                            waypoint = Waypoint(
                                name=wp_data['name'],
                                x=wp_data['x'],
                                y=wp_data['y'],
                                theta=wp_data.get('theta', 0.0),
                                map_name=wp_data.get('map_name', 'map'),
                                connections=wp_data.get('connections', [])
                            )
                            route.waypoints.append(waypoint)
                    
                    self.routes[route.route_id] = route
                    
                self.get_logger().info(f"Loaded {len(self.routes)} routes")
            
            return True
            
        except Exception as e:
            self.get_logger().error(f"Failed to load routes: {e}")
            return False
    
    def load_map_info(self, map_yaml_file: str) -> bool:
        """Load map information from YAML file."""
        try:
            self.get_logger().info(f"Loading map info from: {map_yaml_file}")
            
            with open(map_yaml_file, 'r') as f:
                map_config = yaml.safe_load(f)
            
            # Update map info
            self.map_info.update({
                'resolution': map_config.get('resolution', 0.05),
                'origin': map_config.get('origin', [0.0, 0.0, 0.0]),
                'map_name': map_config.get('image', ''),
                'yaml_file': map_yaml_file
            })
            
            # Load PGM file
            pgm_path = map_config.get('image', '')
            if not pgm_path.startswith('/'):
                # Relative path - combine with YAML directory
                import os
                yaml_dir = os.path.dirname(map_yaml_file)
                pgm_path = os.path.join(yaml_dir, pgm_path)
            
            success = self._load_pgm_map(pgm_path)
            if success:
                self.map_info['loaded'] = True
                self.get_logger().info("Map loaded successfully")
            
            return success
            
        except Exception as e:
            self.get_logger().error(f"Failed to load map info: {e}")
            return False
    
    def _load_pgm_map(self, pgm_file_path: str) -> bool:
        """Load PGM map file."""
        try:
            # Load image using OpenCV
            map_image = cv2.imread(pgm_file_path, cv2.IMREAD_GRAYSCALE)
            
            if map_image is None:
                self.get_logger().error(f"Could not load PGM file: {pgm_file_path}")
                return False
            
            # Create occupancy grid
            self.occupancy_grid = OccupancyGrid()
            self.occupancy_grid.header.frame_id = "map"
            self.occupancy_grid.header.stamp = self.get_clock().now().to_msg()
            
            # Set map metadata
            self.occupancy_grid.info.resolution = self.map_info['resolution']
            self.occupancy_grid.info.width = map_image.shape[1]
            self.occupancy_grid.info.height = map_image.shape[0]
            
            # Set origin
            self.occupancy_grid.info.origin.position.x = self.map_info['origin'][0]
            self.occupancy_grid.info.origin.position.y = self.map_info['origin'][1]
            self.occupancy_grid.info.origin.position.z = self.map_info['origin'][2]
            self.occupancy_grid.info.origin.orientation.w = 1.0
            
            # Convert image to occupancy data
            occupancy_data = []
            for y in range(map_image.shape[0]):
                for x in range(map_image.shape[1]):
                    # Flip Y axis and convert pixel values
                    pixel = map_image[map_image.shape[0] - 1 - y, x]
                    
                    if pixel > 250:  # Free space (white)
                        occupancy_data.append(0)
                    elif pixel < 5:  # Occupied space (black)
                        occupancy_data.append(100)
                    else:  # Unknown space (gray)
                        occupancy_data.append(-1)
            
            self.occupancy_grid.data = occupancy_data
            
            # Update map info
            self.map_info.update({
                'width': map_image.shape[1],
                'height': map_image.shape[0]
            })
            
            self.get_logger().info(
                f"Loaded PGM map: {map_image.shape[1]}x{map_image.shape[0]}, "
                f"resolution: {self.map_info['resolution']}")
            
            return True
            
        except Exception as e:
            self.get_logger().error(f"Failed to load PGM map: {e}")
            return False
    
    def get_waypoint(self, waypoint_name: str) -> Optional[Waypoint]:
        """Get waypoint by name."""
        return self.waypoints.get(waypoint_name)
    
    def get_route(self, route_id: str) -> Optional[Route]:
        """Get route by ID."""
        return self.routes.get(route_id)
    
    def get_all_waypoints(self) -> Dict[str, Waypoint]:
        """Get all waypoints."""
        return self.waypoints.copy()
    
    def get_all_routes(self) -> Dict[str, Route]:
        """Get all routes."""
        return self.routes.copy()
    
    def waypoint_to_pose(self, waypoint: Waypoint) -> PoseStamped:
        """Convert waypoint to PoseStamped."""
        pose = PoseStamped()
        pose.header.frame_id = waypoint.map_name
        pose.header.stamp = self.get_clock().now().to_msg()
        
        pose.pose.position.x = waypoint.x
        pose.pose.position.y = waypoint.y
        pose.pose.position.z = 0.0
        
        # Convert theta to quaternion
        cos_half = math.cos(waypoint.theta / 2.0)
        sin_half = math.sin(waypoint.theta / 2.0)
        
        pose.pose.orientation.x = 0.0
        pose.pose.orientation.y = 0.0
        pose.pose.orientation.z = sin_half
        pose.pose.orientation.w = cos_half
        
        return pose
    
    def find_nearest_waypoint(
        self, 
        pose: PoseStamped, 
        max_distance: float = 2.0
    ) -> Optional[Waypoint]:
        """Find nearest waypoint to given pose."""
        nearest_waypoint = None
        min_distance = max_distance
        
        for waypoint in self.waypoints.values():
            distance = self._calculate_distance(
                pose.pose.position, 
                Point(x=waypoint.x, y=waypoint.y, z=0.0)
            )
            
            if distance < min_distance:
                min_distance = distance
                nearest_waypoint = waypoint
        
        return nearest_waypoint
    
    def find_route(self, start_waypoint: str, end_waypoint: str) -> Optional[Route]:
        """Find route between two waypoints."""
        for route in self.routes.values():
            if (route.start_waypoint == start_waypoint and 
                route.end_waypoint == end_waypoint):
                return route
        return None
    
    def map_to_world(self, map_x: float, map_y: float) -> Point:
        """Convert map coordinates to world coordinates."""
        world_point = Point()
        
        if self.map_info['loaded']:
            world_point.x = (self.map_info['origin'][0] + 
                           map_x * self.map_info['resolution'])
            world_point.y = (self.map_info['origin'][1] + 
                           map_y * self.map_info['resolution'])
            world_point.z = 0.0
        else:
            world_point.x = map_x
            world_point.y = map_y
            world_point.z = 0.0
        
        return world_point
    
    def world_to_map(self, world_point: Point) -> Tuple[float, float]:
        """Convert world coordinates to map coordinates."""
        if self.map_info['loaded']:
            map_x = ((world_point.x - self.map_info['origin'][0]) / 
                    self.map_info['resolution'])
            map_y = ((world_point.y - self.map_info['origin'][1]) / 
                    self.map_info['resolution'])
            return map_x, map_y
        else:
            return world_point.x, world_point.y
    
    def validate_routes(self) -> bool:
        """Validate route connectivity."""
        valid = True
        
        # Check waypoint connections
        for waypoint_name, waypoint in self.waypoints.items():
            for connection in waypoint.connections:
                if connection not in self.waypoints:
                    self.get_logger().error(
                        f"Waypoint '{waypoint_name}' has invalid connection to '{connection}'")
                    valid = False
        
        # Check route waypoints
        for route_id, route in self.routes.items():
            if route.start_waypoint and route.start_waypoint not in self.waypoints:
                self.get_logger().error(
                    f"Route '{route_id}' has invalid start waypoint: '{route.start_waypoint}'")
                valid = False
            
            if route.end_waypoint and route.end_waypoint not in self.waypoints:
                self.get_logger().error(
                    f"Route '{route_id}' has invalid end waypoint: '{route.end_waypoint}'")
                valid = False
        
        if valid:
            self.get_logger().info("Route validation passed")
        else:
            self.get_logger().warning("Route validation failed")
        
        return valid
    
    def export_routes_to_json(self) -> str:
        """Export routes to JSON string."""
        data = {
            'waypoints': [],
            'routes': []
        }
        
        # Export waypoints
        for waypoint in self.waypoints.values():
            data['waypoints'].append({
                'name': waypoint.name,
                'x': waypoint.x,
                'y': waypoint.y,
                'theta': waypoint.theta,
                'map_name': waypoint.map_name,
                'connections': waypoint.connections
            })
        
        # Export routes
        for route in self.routes.values():
            route_data = {
                'route_id': route.route_id,
                'name': route.name,
                'description': route.description,
                'start_waypoint': route.start_waypoint,
                'end_waypoint': route.end_waypoint,
                'estimated_duration': route.estimated_duration
            }
            
            if route.waypoints:
                route_data['waypoints'] = []
                for waypoint in route.waypoints:
                    route_data['waypoints'].append({
                        'name': waypoint.name,
                        'x': waypoint.x,
                        'y': waypoint.y,
                        'theta': waypoint.theta,
                        'map_name': waypoint.map_name,
                        'connections': waypoint.connections
                    })
            
            data['routes'].append(route_data)
        
        return json.dumps(data, indent=2)
    
    def get_map_info(self) -> Dict[str, Any]:
        """Get map information."""
        return self.map_info.copy()
    
    def _calculate_distance(self, p1: Point, p2: Point) -> float:
        """Calculate Euclidean distance between two points."""
        dx = p1.x - p2.x
        dy = p1.y - p2.y
        dz = p1.z - p2.z
        return math.sqrt(dx*dx + dy*dy + dz*dz)
    
    def _publish_map(self):
        """Publish map for visualization."""
        if (self.occupancy_grid and 
            self.map_info['loaded'] and 
            self.map_pub.get_subscription_count() > 0):
            
            self.occupancy_grid.header.stamp = self.get_clock().now().to_msg()
            self.map_pub.publish(self.occupancy_grid)